validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/BlockFactory.java

```diff
@@ -227,57 +256,39 @@ public BeaconBlock createUnsignedBlock(
     final ExecutionPayloadUtil executionPayloadUtil =
         spec.atSlot(slot).getExecutionPayloadUtil().orElseThrow();
-    final MergeTransitionHelpers mergeTransitionHelpers =
-        spec.atSlot(slot).getMergeTransitionHelpers().orElseThrow();
-    final TransitionStore transitionStore = spec.atSlot(slot).getTransitionStore().orElseThrow();
 
-    if (!mergeTransitionHelpers.isMergeComplete(state)) {
-      PowBlock powHead = mergeTransitionHelpers.getPowChainHead(executionEngineChannel);
-      if (!mergeTransitionHelpers.isValidTerminalPowBlock(powHead, transitionStore)) {
-        // Pre-merge, empty payload
-        LOG.info(
-            ColorConsolePrinter.print(
-                String.format(
-                    "Produce pre-merge block: pow_block.total_difficulty(%d) < transition_total_difficulty(%d), PoW blocks left ~%d",
-                    powHead.totalDifficulty.toBigInteger(),
-                    transitionStore.getTransitionTotalDifficulty().toBigInteger(),
-                    transitionStore
-                        .getTransitionTotalDifficulty()
-                        .subtract(powHead.totalDifficulty)
-                        .divide(powHead.difficulty)
-                        .add(UInt256.ONE)
-                        .toBigInteger()),
-                Color.CYAN));
-        return new ExecutionPayload();
-      } else {
-        // Signify merge via producing on top of the last PoW block
-        LOG.info(
-            ColorConsolePrinter.print(
-                String.format(
-                    "Produce transition block: pow_block.total_difficulty(%d) >= transition_total_difficulty(%d)",
-                    powHead.totalDifficulty.toBigInteger(),
-                    transitionStore.getTransitionTotalDifficulty().toBigInteger()),
-                Color.YELLOW));
+    return getExecutionPayloadRefFromSlot(slot)
+        .map(
+            ref ->
+                validateExecutionPayload(
+                    ref.getParentHash(),
+                    executionPayloadUtil.getExecutionPayload(
+                        executionEngineChannel, ref.getPreparePayloadId())))
+        .orElseGet(ExecutionPayload::new);
+  }
 
-        return validateExecutionPayload(
-            powHead.blockHash,
-            executionPayloadUtil.getExecutionPayload(
-                executionEngineChannel, getExecutionPayloadIdFromSlot(slot)));
-      }
+  private Optional<PowBlock> getPowBlockAtTotalDifficulty(
+      UInt256 totalDifficulty, PowBlock head, MergeTransitionHelpers mergeTransitionHelpers) {
+
+    if (head.totalDifficulty.compareTo(totalDifficulty) < 0) {
+      return Optional.empty();
     }
 
-    // Post-merge, normal payload
-    return validateExecutionPayload(
-        state.getLatest_execution_payload_header().getBlock_hash(),
-        executionPayloadUtil.getExecutionPayload(
-            executionEngineChannel, getExecutionPayloadIdFromSlot(slot)));
+    PowBlock block = head;
+    while (true) {
```

Comment:
potential infinite loop 🤔 , could make sense to put a limit on how far we go?

Reply:
I believe I came up with a better variant: a0c4a60

Reply:
looks better but iterations still depend only on external source. Probably irrelevant in practice.